lib/src/source.dart

```diff
@@ -161,15 +161,20 @@ abstract class BoundSource {
   /// Sources should not override this. Instead, they implement [doGetVersions].
   ///
   /// If [maxAge] is given answers can be taken from cache - up to that age old.
-  Future<List<PackageId>> getVersions(PackageRef ref, {Duration maxAge}) {
+  ///
+  /// The [allowedRetractedVersion], if given, is the only version which can be
+  /// selected even if it is marked as retracted. Otherwise, all the returned
+  /// IDs correspond to non-retracted versions.
+  Future<List<PackageId>> getVersions(PackageRef ref,
+      {Duration maxAge, Version allowedRetractedVersion}) {
```

Comment:
Maybe this should be 
```suggestion
      {Duration maxAge, bool allowedRetractedVersions = false}) {
```
To not let the solver logic affect the version listing

Reply:
As we discussed, we will not do this, but keep the `Version allowedRetractedVersion` argument.